File: ios/SsdidWallet/Domain/Model/DidDocument.swift
import Foundation

struct DidDocument: Codable {
    let context: [String]
    let id: String
    var controller: String = ""
    let verificationMethod: [VerificationMethod]
    var authentication: [String] = []
    var assertionMethod: [String] = []
    var capabilityInvocation: [String] = []
    var keyAgreement: [String] = []
    var service: [Service] = []
    var nextKeyHash: String? = nil

    enum CodingKeys: String, CodingKey {
        case context = "@context"
        case id, controller, verificationMethod
        case authentication, assertionMethod, capabilityInvocation
        case keyAgreement, service, nextKeyHash
    }

    init(
        context: [String] = ["https://www.w3.org/ns/did/v1"],
        id: String,
        controller: String = "",
        verificationMethod: [VerificationMethod],
        authentication: [String] = [],
        assertionMethod: [String] = [],
        capabilityInvocation: [String] = [],
        keyAgreement: [String] = [],
        service: [Service] = [],
        nextKeyHash: String? = nil
    ) {
        self.context = context
        self.id = id
        self.controller = controller
        self.verificationMethod = verificationMethod
        self.authentication = authentication
        self.assertionMethod = assertionMethod
        self.capabilityInvocation = capabilityInvocation
        self.keyAgreement = keyAgreement
        self.service = service
        self.nextKeyHash = nextKeyHash
    }

    static func build(
        did: Did,
        keyId: String,
        algorithm: Algorithm,
        publicKeyMultibase: String
    ) -> DidDocument {
        DidDocument(
            id: did.value,
            controller: did.value,
            verificationMethod: [
                VerificationMethod(
                    id: keyId,
                    type: algorithm.w3cType,
                    controller: did.value,
                    publicKeyMultibase: publicKeyMultibase
                )
            ],
            authentication: [keyId],
            assertionMethod: [keyId],
            capabilityInvocation: [keyId]
        )
    }
}

struct Service: Codable {
    let id: String
    let type: String
    let serviceEndpoint: String
}

File: android/app/src/main/java/my/ssdid/wallet/domain/model/DidDocument.kt
package my.ssdid.wallet.domain.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

@Serializable
data class DidDocument(
    @SerialName("@context") val context: List<String> = listOf("https://www.w3.org/ns/did/v1"),
    val id: String,
    val controller: String = "",
    val verificationMethod: List<VerificationMethod>,
    val authentication: List<String> = emptyList(),
    val assertionMethod: List<String> = emptyList(),
    val capabilityInvocation: List<String> = emptyList(),
    val keyAgreement: List<String> = emptyList(),
    val service: List<Service> = emptyList(),
    val nextKeyHash: String? = null
) {
    companion object {
        fun build(did: Did, keyId: String, algorithm: Algorithm, publicKeyMultibase: String): DidDocument {
            return DidDocument(
                id = did.value,
                controller = did.value,
                verificationMethod = listOf(
                    VerificationMethod(
                        id = keyId,
                        type = algorithm.w3cType,
                        controller = did.value,
                        publicKeyMultibase = publicKeyMultibase
                    )
                ),
                authentication = listOf(keyId),
                assertionMethod = listOf(keyId),
                capabilityInvocation = listOf(keyId)
            )
        }
    }
}

@Serializable
data class VerificationMethod(
    val id: String,
    val type: String,
    val controller: String,
    val publicKeyMultibase: String = "",
    val publicKeyJwk: JsonObject? = null
)

@Serializable
data class Service(
    val id: String,
    val type: String,
    val serviceEndpoint: String
)
